Implement "looped" signal for AudioStreamPlayer, 2D and 3D

vaartis opened the pull request:
Track the number of loops on the player and if it changes, emit a singal. 

Closes godotengine/godot-proposals#1641

<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.
-->

TokageItLab reviewed (changes requested):
These codes looks messed up in totality.

It makes no sense to count the number of loops on the signal side. Also, it may cause an implicit overflow. Signal should only notify the fact that a loop has occurred and should be counted on the user script side.

Moreover, the looped signal should fire only when a loop occurs, so should not fire at the same time as finished.

TokageItLab commented:
Ah, I misunderstood a bit. The looped count is something that already exists, and the user determines the upper limit, right? But I think there may be a cleaner way to fire signal.

Also I remember there was some problem with the AnimationPlayer looped flag and I didn't implement it. I'll take a look at it later...

vaartis commented:
> It makes no sense to count the number of loops on the signal side. Also, it may cause an implicit overflow. Signal should only notify the fact that a loop has occurred and should be counted on the user script side.

The overflow here does not matter, even if it overflows the number still changes and this is what matters to fire the signal. I just used the type that get_loop_count() returns.

> Moreover, the looped signal should fire only when a loop occurs, so should not fire at the same time as finished.

It only fires if the stream _did not_ end, hence the else if on the condition that checks if it ended.

TokageItLab commented on scene/2d/audio_stream_player_2d.h:
```diff
@@ -52,6 +52,7 @@ class AudioStreamPlayer2D : public Node2D {
 	};
 
 	Vector<Ref<AudioStreamPlayback>> stream_playbacks;
+	HashMap<Ref<AudioStreamPlayback>, int> stream_looped;
```
It is obviously hackey to have a map of each stream just to subscribe to each loop. So you should implement a looped flag in the AudioStreamPlayback, then connect and disconnect to the AudioStreamPlayer correctly and then fire the signal there.

vaartis replied:
Should the looped flag be reset when it's read by the playback or by the player when it's read? I suppose it's probably a better idea to auto-reset it, but also then why does get_loop_count() even exist? It's not actually available from GDScript, despite being in the docs you cannot call it.

vaartis replied:
Actually, if you meant "signal" by "flag" then this is probably a better solution, yes.

TokageItLab replied:
I'm not sure what you mean by reset. It just needs to fire a signal from AudioStreamPlayback at the moment of loop, i.e. the moment the `loops` increments in the AudioStreamPlayback child class. Then subscribes to it in AudioStreamPlayer, and then fires more signal from AudioStreamPlayback. Signals must be connected and disconnected at the time of loading and deleting the streams.

TokageItLab replied:
> So you should implement a looped flag

 Oh, sorry, it was signal, not flag.

vaartis replied:
Re-implemented this with signaling and tracking the looped on the players with `CONNECT_REFERENCE_COUNTED`, since this seems to be one of the cases where this is what we want.

I hope my implementation of the `looped` signal on the randomizer is OK, I can think of a use case for this signal on it so I thought I'd add it. The playbacks that I didn't add the signal to do not support looping and always return 0 for the count of loops.

TokageItLab commented on scene/animation/animation_player.cpp:
```diff
@@ -1064,6 +1064,10 @@ void AnimationPlayer::_animation_process_data(PlaybackData &cd, double p_delta,
 			end_notify = prev_pos > 0; // Notify only if not already at the beginning.
 		}
 	}
+
+	if (looped_flag != Animation::LOOPED_FLAG_NONE) {
+		emit_signal(SceneStringNames::get_singleton()->animation_looped, playback.assigned);
+	}
```
I faintly remember that the reason we did not implement this in #70278 before was that depending on the length of the animation, this might not fire correctly.

And I remember that these flags were only "populated" for the purpose of calculating looped time, and that the flags were not always fired correctly during the loop. Implementing this would require more strict testing, so if you want to focus on audio for now, this should be removed and a separate PR should be used.

vaartis replied:
I'll probably just remove it for now, then.

vaartis committed:
Implement "looped" signal for AudioStreamPlayer, 2D and 3D

Emit a signal from a playback whenever a loop occurs, the player
connects to that signal and forwards it.

reduz commented:
I oppose the merging of this PR on the following grounds:

* Emitting a signal from the audio thread is wrong in many accounts 1) Audio thread must never run complex logic that can allocate memory or run game logic as a result of a signal, because its time (callback) constrained. 2) A signal emitted from the audio thread is useless and dangerous.
* You can already query the loop count, so all you have to do is check every frame whether the audio has looped.

TokageItLab commented:
This PR originally tried to fire a signal by storing the loop counts of all streams in a map, is this the correct way to do it? Considering the many possibilities of not using the signal, it seemed to me to be expensive.

Perhaps the correct way to do it is to set up a script that monitors only certain Streams when a signal is needed on the user's side.

vaartis commented:
> Perhaps the correct way to do it is to set up a script that monitors only certain Streams when a signal is needed on the user's side.
Is there a way to check if a signal is needed at all? And if so, how should it track the looping? If emitting a signal on the audio thread is not a good idea, maybe just using the old approach with the dictionary is fine, as long as I can somehow know if I need to track only certain streams.
